src/MonitorProfiler/ProbeInstrumentation/ProbeInstrumentation.cpp

```diff
@@ -0,0 +1,313 @@
+    //
+    //
+
+    //
+    // This method un-flattens the passed in data, reconstructing it into an easier-to-understand format
+    // before passing off the request to the worker thread.
+    //
+
+    vector<UNPROCESSED_INSTRUMENTATION_REQUEST> requests;
+    requests.reserve(count);
+
+    ULONG32 offset = 0;
+    for (ULONG32 i = 0; i < count; i++)
+    {
+        if (UINT32_MAX - offset < argumentCounts[i])
+        {
+            return E_INVALIDARG;
+        }
+
+        vector<ULONG32> tokens;
+        tokens.reserve(argumentCounts[i]);
+        for (ULONG32 j = 0; j < argumentCounts[i]; j++)
+        {
+            tokens.push_back(argumentBoxingTypes[offset+j]);
+        }
+        offset += argumentCounts[i];
+
+        UNPROCESSED_INSTRUMENTATION_REQUEST request;
+        request.functionId = static_cast<FunctionID>(functionIds[i]);
+        request.boxingTypes = tokens;
+
+        requests.push_back(request);
+    }
+
+    m_probeManagementQueue.Enqueue({ProbeWorkerInstruction::INSTALL_PROBES, requests});
+
+    return S_OK;
+}
+
+HRESULT ProbeInstrumentation::RequestFunctionProbeUninstallation()
+{
+    m_pLogger->Log(LogLevel::Debug, _LS("Probe uninstallation requested"));
+
+    if (!HasRegisteredProbe())
+    {
+        return S_FALSE;
+    }
+
+    PROBE_WORKER_PAYLOAD payload = {};
+    payload.instruction = ProbeWorkerInstruction::UNINSTALL_PROBES;
+    m_probeManagementQueue.Enqueue(payload);
+
+    return S_OK;
+}
+
+bool ProbeInstrumentation::HasRegisteredProbe()
+{
+    return m_probeFunctionId != 0;
+}
+
+HRESULT ProbeInstrumentation::InstallProbes(vector<UNPROCESSED_INSTRUMENTATION_REQUEST>& requests)
+{
+    HRESULT hr;
+
+    lock_guard<mutex> lock(m_instrumentationProcessingMutex);
+
+    if (!HasRegisteredProbe() ||
+        AreProbesInstalled())
+    {
+        return E_FAIL;
+    }
+
+    unordered_map<pair<ModuleID, mdMethodDef>, INSTRUMENTATION_REQUEST, PairHash<ModuleID, mdMethodDef>> newRequests;
+
+    vector<ModuleID> requestedModuleIds;
+    vector<mdMethodDef> requestedMethodDefs;
+
+    // JSFIX: Handle OOM scenarios.
+    requestedModuleIds.reserve(requests.size());
+    requestedMethodDefs.reserve(requests.size());
+
+    for (auto const& req : requests)
+    {
+        INSTRUMENTATION_REQUEST processedRequest;
+
+        // For now just use the function id as the uniquifier.
+        // Consider allowing the caller to specify one.
+        processedRequest.uniquifier = static_cast<ULONG64>(req.functionId);
+        processedRequest.boxingTypes = req.boxingTypes;
+
+        IfFailLogRet(m_pCorProfilerInfo->GetFunctionInfo2(
+            req.functionId,
+            NULL,
+            nullptr,
+            &processedRequest.moduleId,
+            &processedRequest.methodDef,
+            0,
+            nullptr,
+            nullptr));
+
+        IfFailLogRet(m_pAssemblyProbePrep->PrepareAssemblyForProbes(processedRequest.moduleId));
```

Comment:
Not something for this review, but it would probably be worth testing this system with AssemblyLoadContext scenarios